src/core/eme/create_session.ts

```diff
@@ -0,0 +1,191 @@
+  initData: Uint8Array,
+  initDataType: string,
+  session: MediaKeySession|IMediaKeySession
+) : Observable<IGetSessionEvent> {
+  log.debug("eme: load persisted session", storedSessionId);
+
+  return castToObservable(session.load(storedSessionId))
+    .catch((error) => {
+      log.warn("eme: no data stored for the loaded session.",
+        storedSessionId);
+
+      const loadedSession = $loadedSessions.get(initData, initDataType);
+      if (loadedSession != null) {
+        $loadedSessions.closeSession(loadedSession);
+      }
+      $storedSessions.delete(initData);
+
+      throw error;
+    })
+    .map((success) => {
+      if (success) {
+        $loadedSessions.add(initData, initDataType, session);
+        return {
+          type: "loaded-persistent-session" as "loaded-persistent-session",
+          value: { session },
+        };
+      } else {
+        return {
+          type: "created-session" as "created-session",
+          value: { session },
+        };
+      }
+    }).do(() => $storedSessions.add(initData, session));
+}
+
+/**
+ * Create a new Session on the given MediaKeys, corresponding to the given
+ * initializationData.
+ * If session creating fails, remove the oldest MediaKeySession loaded and
+ * retry.
+ *
+ * /!\ This only creates new sessions.
+ * It will fail if $loadedSessions already has a MediaKeySession with
+ * the given initializationData.
+ * @param {Uint8Array} initData
+ * @param {string} initDataType
+ * @param {Object} mediaKeysInfos
+ */
+function createSession(
+  initData: Uint8Array,
+  initDataType: string,
+  mediaKeysInfos: IMediaKeysInfos
+) : Observable<IGetSessionEvent> {
+
+  const {
+    keySystemOptions,
+    keySystemAccess,
+    mediaKeys,
+  } = mediaKeysInfos;
+  if (mediaKeys.createSession == null) {
+    throw new Error("Invalid MediaKeys implementation: Missing createSession");
+  }
+
+  const mksConfig = keySystemAccess.getConfiguration();
+  const sessionTypes = mksConfig.sessionTypes;
+  const hasPersistence = (
+    sessionTypes && arrayIncludes(sessionTypes, "persistent-license")
+  );
+  const sessionType = hasPersistence && keySystemOptions.persistentLicense ?
+    "persistent-license" : "temporary";
+
+  log.debug(`eme: create a new ${sessionType} session`);
+
+  // TODO TS bug? I don't get the problem here.
+  const session : IMediaKeySession|MediaKeySession =
+    (mediaKeys as any).createSession(sessionType);
+
+  $loadedSessions.add(initData, initDataType, session);
+
+  if (hasPersistence && keySystemOptions.persistentLicense) {
+    // if a persisted session exists in the store associated to this initData,
+    // we reuse it without a new license request through the `load` method.
+    const storedEntry = $storedSessions.get(initData);
+    if (storedEntry) {
+      return loadPersistentSession(
+        storedEntry.sessionId, initData, initDataType, session
+      );
+    }
+  }
+  return Observable.of({
+    type: "created-session" as "created-session",
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Remove this unnecessary cast. ([typescript:S4325](https://sonarcloud.io/organizations/rx-player/rules#rule_key=typescript%3AS4325))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=rx-player&pullRequest=187&issues=AWMmgcqUmtQEh111i0u4&open=AWMmgcqUmtQEh111i0u4)

[ISSUE_KEY]: AWMmgcqUmtQEh111i0u4